### PR TITLE
Fix underlying connection not being closed on protocol error

### PR DESCRIPTION
**Pull request by samherrmann**
Please expand the commit messages for details.

### CODE REVIEW

File: jsonrpc2.go
Comment (samherrmann) [inline jsonrpc2.go:698]: This is the source of the bug that's being fixed. Before, `close(c.disconnect)` was being called here but `c.stream.Close()` was not. `c.stream.Close()` was only being called if `Conn` was explicitly being closed by `Conn.Close()`.

I rearranged the code so that the "closing" procedure is the same regardless if the cause is an error in `Conn.readMessages` or if the cause is a call to `Conn.Close()`. This should ensure that the state of the underlying connection (i.e. `c.stream`) matches what `Conn.DiconnectNotify()` reports.

From a high level, this is what it looks like:

```go
func (c *Conn) readMessages() {
   var err error
   for err == nil {
     // read from underlying connection
   }
   c.close(err)
}

func (c *Conn) Close() error {
   return c.close(nil)
}

func (c *Conn) close(cause err) error {
   // common closing procedure
}

```

**Comment by samherrmann** [comment]
Let me know if you wish to keep the change set of this PR to a minimum and remove the second commit.

File: jsonrpc2.go
Comment (samherrmann) [inline jsonrpc2.go:371]: Since there is now a single closing procedure, there doesn't appear to be the need for a `closing` AND `shutdown` state any more.

File: jsonrpc2_test.go
Comment (samherrmann) [inline jsonrpc2_test.go:321]: I switched from `jsonrpc2.NewBufferedStream(conn, jsonrpc2.VarintObjectCodec{}), nil)` to `jsonrpc2.NewPlainObjectStream(conn)` solely because the latter looks cleaner/shorter. The type of stream being used shouldn't impact the results of the these tests. I am happy to switching back to the former though if you prefer.

File: jsonrpc2_test.go
Comment (samherrmann) [inline jsonrpc2_test.go:319]: Instead of modifying `readWriteCloser` so that it would allow us to test if it's been closed, I just used `net.Pipe()` gtom std lib instead.

File: jsonrpc2.go
Comment (samherrmann) [inline jsonrpc2.go:699]: This line is the source of the bug that's being fixed. Before, `close(c.disconnect)` was being called here but `c.stream.Close()` was not. `c.stream.Close()` was only being called if `Conn` was explicitly being closed by `Conn.Close()`.

I rearranged the code so that the "closing" procedure is the same regardless if the cause is an error in `Conn.readMessages` or if the cause is a call to `Conn.Close()`. This should ensure that the state of the underlying connection matches what `Conn.DiconnectNotify()` reports.

From a high level, this is what it looks like:

```go
func (c *Conn) readMessages() {
   var err error
   for err == nil {
     // read from underlying connection
   }
   // for-loop exits when an error is encountered. Start closing the connection....
   c.close(err)
}

func (c *Conn) Close() error {
   return c.close(nil)
}

func (c *Conn) close(cause err) error {
   // Common closing procedure.
}

```

**Review by keegancsmith — APPROVED** [review]
LGTM! One inline comment, if it looks good to you I'll go ahead and land this :)

File: jsonrpc2.go
Comment (samherrmann) [reply jsonrpc2.go]: Agreed. Addressed in 0c9de81282b6d7e6070ef5c7879629f17e062bc6.

File: jsonrpc2.go
Comment (samherrmann) [reply jsonrpc2.go:685]: Excellent observation! I definitely overlooked this before you drew my attention to it. As it turns out, from what I can see, we don't actually need to send `ErrClosed` through the channel because the [channel listener infers `ErrClosed` when the channel is being closed](https://github.com/sourcegraph/jsonrpc2/blob/master/jsonrpc2.go#L552-L555). Based on that, I'm coming to the conclusion that having the following two lines...

```go
call.done <- ErrClosed
close(call.done)
```
...is redundant and we can reduce the code to just `close(call.done)` (see 85075f0782fc4062813f12fc5d41a787c97a44ab). A follow-on question that may arise from my conclusion is "_what about errors other than `ErrClosed`? Now those don't get sent to `c.pending`_". It seems to me that if a connection is being closed, the cause of that closure is not the concern of a pending request. If a pending request is being cancelled due to the connection being closed, then all it cares about is `ErrClosed`.

Now, this investigation made me pay more attention to `TestConn_Close: waiting for response`, which is the test that I believe is responsible for verifying the behavior discussed above. That made me discover an issue with the test, which I discuss [here](https://github.com/sourcegraph/jsonrpc2/pull/64/commits/85075f0782fc4062813f12fc5d41a787c97a44ab#r1099163167).

**Review by keegancsmith — APPROVED** [review]

**Comment by samherrmann** [comment]
@keegancsmith, thank you very much for the quick turnaround time on the review 🙂

**Pull request merged**
